fix(ivy): ngcc - several fixes related to decorators

JoostK opened the pull request:
**refactor(ivy): ngcc - categorize the various decorate calls upfront**

Any decorator information present in TypeScript is emitted into the
generated JavaScript sources by means of `__decorate` call. This call
contains both the decorators as they existed in the original source
code, together with calls to `tslib` helpers that convey additional
information on e.g. type information and parameter decorators. These
different kinds of decorator calls were not previously distinguished on
their own, but instead all treated as `Decorator` by themselves. The
"decorators" that were actually `tslib` helper calls were conveniently
filtered out because they were not imported from `@angular/core`, a
characteristic that ngcc uses to drop certain decorators.

Note that this posed an inconsistency in ngcc when it processes
`@angular/core`'s UMD bundle, as the `tslib` helper functions have been
inlined in said bundle. Because of the inlining, the `tslib` helpers
appear to be from `@angular/core`, so ngcc would fail to drop those
apparent "decorators". This inconsistency does not currently cause any
issues, as ngtsc is specifically looking for decorators based on  their
name and any remaining decorators are simply ignored.

This commit rewrites the decorator analysis of a class to occur all in a
single phase, instead of all throughout the `ReflectionHost`. This
allows to categorize the various decorate calls in a single sweep,
instead of constantly needing to filter out undesired decorate calls on
the go. As an added benefit, the computed decorator information is now
cached per class, such that subsequent reflection queries that need
decorator information can reuse the cached info.

---

**fix(ivy): ngcc - recognize suffixed tslib helpers**

An identifier may become repeated when bundling multiple source files
into a single bundle, so bundlers have a strategy of suffixing non-unique
identifiers with a suffix like $2. Since ngcc operates on such bundles,
it needs to process potentially suffixed identifiers in their canonical
form without the suffix. The "ngx-pagination" package was previously not
compiled fully, as most decorators were not recognized.

This commit ensures that identifiers are first canonicalized by removing
the suffix, such that they are properly recognized and processed by ngcc.

Fixes #31540

---

**fix(ivy): ngcc - render decorators in UMD and CommonJS bundles correctly**

In #31426 a fix was implemented to render namespaced decorator imports
correctly, however it turns out that the fix only worked when decorator
information was extracted from static properties, not when using 
`__decorate` calls.

This commit fixes the issue by creating the decorator metadata with the
full decorator expression, instead of only its name.
 
Closes #31394

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -810,7 +810,7 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
       const decoratorIdentifier = decoratorExpression;
       return {
-        name: decoratorIdentifier.text,
+        name: stripDollarSuffix(decoratorIdentifier.text),
```
Not sure why this is needed?

The `decoratorIdentifier.text` here should be something like `Directive`, `Component`, etc. Surely it will never be `Directive$1`?

JoostK replied:
@petebacondarwin So I looked into this, and the way we deal with decorators is a bit strange (as you suspected). Given a piece of code as follows:

```javascript
SomeDirective = tslib_1.__decorate([
      Directive({ selector: '[someDirective]' }),
      tslib_1.__param(2, Inject(INJECTED_TOKEN)),
      tslib_1.__metadata("design:paramtypes", [ViewContainerRef,
          TemplateRef, String])
  ], SomeDirective);
```

All of the `Directive(...)`, `tslib_1.__param(...)` and `tslib_1.__metadata(...)` calls are handled in this function and converted to a `Decorator` object. This is why the `stripDollarSuffix` helps us out with stuff like `__param$2` here, but I certainly agree that this is unexpected. It seems strange to be reflecting thse `__param` and `__metadata` calls to be actual `Decorator` instances, as they really weren't in the original source code.

The reason this does not cause issues right now is because we filter out any `Decorator` that is _not_ from `@angular/core`, e.g. here (and all other places that deal with decorators):

https://github.com/angular/angular/blob/cb848b94102855d1e773892b25375304102cab69/packages/compiler-cli/ngcc/src/host/esm2015_host.ts#L575-L585

I would argue that this filter is not desirable, as the host should simply reflect _all_ decorators that were in the original source code. ngtsc will correctly verify the import specifier to ignore any non-Angular decorators by itself, this is not something that ngcc needs to do upfront.

We do currently have an inconsistency for `@angular/core`'s UMD bundle, which contain the tslib helpers inline. In that case, the `__metadata` and `__param` "decorators" don't have an import specifier, so they are believed to be from Angular itself. You can see commit 5669473976582147332826a8c2342ed04e69405b with a workaround, although this does not seem like the correct way to fix it but is rather a hack.

My proposal would be to clean up `reflectDecoratorCall` and allow it to return something along the lines of `Decorator|DecoratorMetadata|ParamDecorator|null` where `DecoratorMetadata` is a special type to represent `__metadata` and `ParamDecorator` corresponds with `__param`. This avoids the mixing of actual decorators as they appeared in the original source, versus the synthetic decorator information as inserted during TypeScript emit.

This will allow us to replace the `isFromCore` filter by a filter that only accepts objects of type `Decorator`.

JoostK committed:
refactor(ivy): ngcc - categorize the various decorate calls upfront

Any decorator information present in TypeScript is emitted into the
generated JavaScript sources by means of `__decorate` call. This call
contains both the decorators as they existed in the original source
code, together with calls to `tslib` helpers that convey additional
information on e.g. type information and parameter decorators. These
different kinds of decorator calls were not previously distinguished on
their own, but instead all treated as `Decorator` by themselves. The
"decorators" that were actually `tslib` helper calls were conveniently
filtered out because they were not imported from `@angular/core`, a
characteristic that ngcc uses to drop certain decorators.

Note that this posed an inconsistency in ngcc when it processes
`@angular/core`'s UMD bundle, as the `tslib` helper functions have been
inlined in said bundle. Because of the inlining, the `tslib` helpers
appear to be from `@angular/core`, so ngcc would fail to drop those
apparent "decorators". This inconsistency does not currently cause any
issues, as ngtsc is specifically looking for decorators based on  their
name and any remaining decorators are simply ignored.

This commit rewrites the decorator analysis of a class to occur all in a
single phase, instead of all throughout the `ReflectionHost`. This
allows to categorize the various decorate calls in a single sweep,
instead of constantly needing to filter out undesired decorate calls on
the go. As an added benefit, the computed decorator information is now
cached per class, such that subsequent reflection queries that need
decorator information can reuse the cached info.

JoostK committed:
fix(ivy): ngcc - recognize suffixed tslib helpers

An identifier may become repeated when bundling multiple source files
into a single bundle, so bundlers have a strategy of suffixing non-unique
identifiers with a suffix like $2. Since ngcc operates on such bundles,
it needs to process potentially suffixed identifiers in their canonical
form without the suffix. The "ngx-pagination" package was previously not
compiled fully, as most decorators were not recognized.

This commit ensures that identifiers are first canonicalized by removing
the suffix, such that they are properly recognized and processed by ngcc.

Fixes #31540

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1365,6 +1392,49 @@ export type ParamInfo = {
+  decorator: Decorator;
+}
+
+export type DecorateCall = DecoratorMetadata | DecoratedParam | DecoratorCall;
```
The `DecorateCall` name is a bit misleading. On a first glance it seemed like it should represent a call to `__decorate`. But I think really it is representing an argument on a call to `__decorate` (and also an argument on a call to `__param` as it turns out!).

Naming is a killer and we could debate this stuff all day. I don't have a very great suggestion other than possibly, `HelperArg`??  (In which case `reflectDecorateCall()` would be renamed to `reflectHelperArg()` and `reflectDecorateCallsFromHelperCall()` would become `reflectDecorateHelperArgs()`?

BTW I wonder why we use discriminated interfaces here. Is there a benefit over using real classes? Is it just a coding style or are there more concrete benefits?

JoostK replied:
I wasn't too fond of it either. I like your proposal of `HelperArg`, or maybe more specific `DecorateHelperArg`. I will play with the naming scheme some more.

The usage of discriminated interfaces here is just a coding style. I tend to not write classes/depend on `instanceof` checks. The main benefit is that TypeScript knows all possible options, so that narrowing and completeness checks work really accurately.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1365,6 +1392,49 @@ export type ParamInfo = {
+export interface DecoratedParam {
+  type: 'param';
+  index: ts.Node;
+  decorator: ts.Node;
```
Or given my other comment, perhaps `helperArg`?

JoostK replied:
Strictly speaking, the argument to `__param` is always a regular decorator, it won't be any of the helpers. The name used here corresponds with the name that is used in the declaration of `__param`:

https://github.com/microsoft/tslib/blob/293765b5820032c836ce75df709194280d57f843/tslib.d.ts#L19

I was wondering if these interfaces should carry the `ts.Node` or interpret the value into e.g. `Decorator`, but decided to just refer to the uninterpreted `ts.Node`.

petebacondarwin replied:
OK, I was misled by the fact that this value is passed through to `reflectDecorateCall()` again which once again checks for the type of this argument:
https://github.com/angular/angular/pull/31614/files#diff-56411fd1f865093a818fbf15a5ea15f6R1209

Perhaps then we should check for this being a `CallExpression` in the code that creates the `DecoratedParam` object and change the type of this property to `DecoratorCall`?

petebacondarwin commented on packages/compiler-cli/ngcc/test/helpers/utils.ts:
```diff
@@ -78,6 +78,22 @@ export function convertToDirectTsLibImport(filesystem: TestFile[]) {
+var __param$2 = null;
+var __extends$2 = null;
+var __assign$2 = null;
+`).replace(/tslib_1\.([_a-z]+)/gi, '$1\$2');
```
Nice! That `$1\$2` really threw me for a second! :-)

gkalpak replied:
Well, actually... 😛 
`\$2` is the same as `$2` in this case (i.e. `\` doesn't do anything here). And the only reason it works is because there is only one captured group (so `$2` doesn't correspond to a captured group, so it is added literally). The proper way to insert a literal `$` is `$$`: `'$1$$2` 🤑

JoostK replied:
@gkalpak wow, I would not have expected that! Thanks for chiming in 😃 I will go with the `$$` 💵 💵

gkalpak replied:
BTW, for future reference, here are [the docs][1] :smiley:

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_string_as_a_parameter

petebacondarwin reviewed (approved):
Great work @JoostK - some minor naming suggestions.

JoostK commented on packages/compiler-cli/ngcc/test/host/commonjs_host_spec.ts:
```diff
@@ -951,23 +951,6 @@ exports.ExternalModule = ExternalModule;
           expect(decorators[0]).toEqual(jasmine.objectContaining({name: 'Directive'}));
         });
 
-        it('should use `getImportOfIdentifier()` to retrieve import info', () => {
```
I decided to delete these tests as the refactor into a single analysis phase affects the interaction with the spy, making this test brittle and not really of added value, IMHO. Assertions around the import location of a decorator have been added to other tests, where necessary.

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -733,102 +779,194 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+
+          if (entry.type === 'decorator') {
+            // The helper arg was reflected to represent an actual decorator
+            if (this.isFromCore(entry.decorator)) {
```
Ideally, this check should not be here. The `ReflectionHost` itself should tell the exact decorators as they occurred in the original source, downstream consumers are then responsible for selecting those from `@angular/core`, if need be.

The reason I left it in, is that the static properties analysis pipeline also has the `isFromCore` filter and removing it there will cause tests to fail, as we'd end up removing _all_ decorators during the render phase, instead of just those from `@angular/core`. If we want to clean this up, it feels like we should perform the `isFromCore` filter when assigning into `AnalyzedClass.decorators`, as that is also the source of how decorators to remove are determined.

petebacondarwin replied:
I think it is fine to leave this as it is. Ngcc should never need to parse non-core decorators.

JoostK committed:
fix(ivy): ngcc - render decorators in UMD and CommonJS bundles correctly

In #31426 a fix was implemented to render namespaced decorator imports
correctly, however it turns out that the fix only worked when decorator
information was extracted from static properties, not when using
`__decorate` calls.

This commit fixes the issue by creating the decorator metadata with the
full decorator expression, instead of only its name.

Closes #31394

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -542,6 +546,71 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * information is computed either from static properties if present, or using `tslib.__decorate`
+   * helper calls otherwise. The computed result is cached per class.
+   *
+   * @param classSymbol
```
This jsdoc tag should either be populated with a description or removed.

gkalpak replied:
The conversation was marked as resolved, but the comment above hasn't been addressed.

petebacondarwin replied:
I think it has https://github.com/angular/angular/pull/31614/files#diff-56411fd1f865093a818fbf15a5ea15f6R554

gkalpak replied:
Ooops, you are right. I forgot I was looking at an individual commit. Sorry for the noise 😇

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
````diff
@@ -1397,27 +1546,55 @@ export function isAssignment(node: ts.Node): node is ts.AssignmentExpression<ts.
+ * __decorate([], SomeDirective.prototype, "member", void 0);
+ * ```
+ *
+ * @param call the call expression that is tested to represent a class decorator call.
````
typo:

@param call the call expression that is tested to represent a ~~class~~member decorator call.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -894,8 +894,8 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
 
-    const name = helperCallFn.text;
-    if (name === '__metadata') {
+    const canonicalName = stripDollarSuffix(helperCallFn.text);
```
I think we can make use of the `getCalleeName()` helper and remove this line and the few lines above. E.g.

instead of 

```ts
const helperCallFn =
    ts.isPropertyAccessExpression(call.expression) ? call.expression.name : call.expression;
if (!ts.isIdentifier(helperCallFn)) {
  return null;
}

const canonicalName = stripDollarSuffix(helperCallFn.text);
```

we could have

```ts
const helperName = getCalleeName(call);
if (helperName === null) {
  return null;
}
```

JoostK replied:
Nice one, we can indeed!

petebacondarwin reviewed (approved):
Awesome PR @JoostK - the first commit is a thing of beauty.
Approved with a few minor nits.

JoostK committed:
fixup! refactor(ivy): ngcc - categorize the various decorate calls upfront

JoostK committed:
fixup! fix(ivy): ngcc - recognize suffixed tslib helpers

gkalpak commented on integration/ngcc/test.sh:
```diff
@@ -68,7 +68,8 @@ grep "_MatMenuBase.ngBaseDef = ɵngcc0.ɵɵdefineBase({ inputs: {" node_modules/
 # Did it handle namespace imported decorators in UMD?
-grep "type: core.Injectable," node_modules/@angular/cdk/bundles/cdk-a11y.umd.js
+grep "type: core.Injectable" node_modules/@angular/common/bundles/common.umd.js # using `__decorate` syntax
+grep "type: core.Injectable," node_modules/@angular/cdk/bundles/cdk-a11y.umd.js # using static properties
```
Since the format of the unprocessed code is important for this test, could we add a check before processing to ensure the two bundles contains the expected code-constructs (to future-proof against changing the format of one of the bundles in the future)?

JoostK replied:
Good idea, done!

gkalpak commented on packages/compiler-cli/ngcc/test/helpers/utils.ts:
```diff
@@ -82,14 +82,14 @@ export function convertToInlineTsLib(filesystem: TestFile[]) {
+var __param$2 = null;
+var __extends$2 = null;
+var __assign$2 = null;
+`).replace(/tslib_1\.([_a-z]+)/gi, '$1$$2');
```
Nit: I would prefer having tests both with and without suffixes (to make sure we handle both cases correctly).

JoostK replied:
I added an additional configuration without suffx

gkalpak reviewed (approved):
💯

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -733,102 +779,194 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+      return null;
+    }
+
+    // We found a decorator!
```
Hahaha, this was in the original code and I found it too nice to remove 😁

JoostK committed:
fixup! fix(ivy): ngcc - recognize suffixed tslib helpers

JoostK committed:
fixup! fix(ivy): ngcc - render decorators in UMD and CommonJS bundles correctly

JoostK commented:
merge-assistance: Angular bot is incorrect, `ci-codefresh-bazel` is succssful.
